libraries/rush-lib/src/logic/operations/CacheableOperationPlugin.ts

```diff
@@ -0,0 +1,1100 @@
+          throw new InternalError('Cobuild cluster id is not defined');
+        }
+        buildCacheContext.cobuildLock = new CobuildLock({
+          cobuildConfiguration,
+          projectBuildCache,
+          cobuildClusterId: buildCacheContext.cobuildClusterId,
+          lockExpireTimeInSeconds: PERIODIC_CALLBACK_INTERVAL_IN_SECONDS * 3,
+          packageName,
+          phaseName
+        });
+      }
+    }
+    return buildCacheContext.cobuildLock;
+  }
+
+  private _getBuildCacheTerminal({
+    record,
+    buildCacheConfiguration,
+    rushProject,
+    logFilenameIdentifier,
+    quietMode,
+    debugMode
+  }: {
+    record: OperationExecutionRecord;
+    buildCacheConfiguration: BuildCacheConfiguration | undefined;
+    rushProject: RushConfigurationProject;
+    logFilenameIdentifier: string;
+    quietMode: boolean;
+    debugMode: boolean;
+  }): ITerminal {
+    const buildCacheContext: IOperationBuildCacheContext =
+      this._getBuildCacheContextByOperationExecutionRecordOrThrow(record);
+    if (!buildCacheContext.buildCacheTerminal) {
+      buildCacheContext.buildCacheTerminal = this._createBuildCacheTerminal({
+        record,
+        buildCacheConfiguration,
+        rushProject,
+        logFilenameIdentifier,
+        quietMode,
+        debugMode
+      });
+    } else if (buildCacheContext.buildCacheProjectLogWritable?.isOpen === false) {
+      // The ProjectLogWritable is closed, re-create one
+      buildCacheContext.buildCacheTerminal = this._createBuildCacheTerminal({
+        record,
+        buildCacheConfiguration,
+        rushProject,
+        logFilenameIdentifier,
+        quietMode,
+        debugMode
+      });
+    }
+
+    return buildCacheContext.buildCacheTerminal;
+  }
+
+  private _createBuildCacheTerminal({
+    record,
+    buildCacheConfiguration,
+    rushProject,
+    logFilenameIdentifier,
+    quietMode,
+    debugMode
+  }: {
+    record: OperationExecutionRecord;
+    buildCacheConfiguration: BuildCacheConfiguration | undefined;
+    rushProject: RushConfigurationProject;
+    logFilenameIdentifier: string;
+    quietMode: boolean;
+    debugMode: boolean;
+  }): ITerminal {
+    const silent: boolean = record.runner.silent;
+    if (silent) {
+      const nullTerminalProvider: NullTerminalProvider = new NullTerminalProvider();
+      return new Terminal(nullTerminalProvider);
+    }
+
+    let cacheConsoleWritable: TerminalWritable;
+    // This creates the writer, only do this if necessary.
+    const collatedWriter: CollatedWriter = record.collatedWriter;
+    const cacheProjectLogWritable: ProjectLogWritable | undefined = this._tryGetBuildCacheProjectLogWritable({
+      record,
+      buildCacheConfiguration,
+      rushProject,
+      collatedTerminal: collatedWriter.terminal,
+      logFilenameIdentifier
+    });
+
+    if (quietMode) {
+      cacheConsoleWritable = new DiscardStdoutTransform({
+        destination: collatedWriter
+      });
```

Comment:
```suggestion
      cacheConsoleWritable = new DiscardStdoutTransform({
        destination: collatedWriter
      });
      
      cacheConsoleWritable = new TextRewriterTransform({
        destination: cacheConsoleWritable,
        normalizeNewlines: NewlineKind.Lf,
        ensureNewlineAtEnd: true
      });
```

This seems to fix the issue. Not entirely sure why.

Reply:
Maybe it simply shuffles a race condition, by changing the `writeChunk()` buffering.  

It's odd that your change would not make the problem worse, since `TextRewriterTransform` can perform additional writes during `close()` whereas `DiscardStdoutTransform` does not buffer anything.

The `autocloseDestination` semantics are identical for both classes.

Reply:
Ian's change here prevents the DiscardStdoutTransform from throwing during a chunk write. Most likely the control flow when that error is thrown breaks the pipeline in an unexpected way that results in keeping the writer open.

Reply:
I believe the issue should be fixed by this commit: [[rush] Implementation of Cobuild feature by chengcyber · Pull Request #3949 · microsoft/rushstack (github.com)](https://github.com/microsoft/rushstack/pull/3949/commits/616dee25ef7cd5e7eb55706d24bf10806eb05026)
which is exactly Ian posted in this thread.


The culprit is not the `operationMetedataManager#saveAsync` method, instead it is caused by `operationMetadataManager#tryRestoreAsync`.

where the terminal is `buildCacheTerminal` i.e. `DiscardStdoutTransform`, and it complains about the unnormalized newlines.

The fix contains:
1. Mark the `cachedRestored` flag before invoking `operationMetadataManager#tryRestoreAsync`, this makes the error from `DiscardStdoutTransform` correctly thrown out.
2. Use the `TextRewriterTransform` pipes chunk text before `DiscardStdoutTransform`, it ensures the `DiscardStdoutTransform` can get normalized newlines when reproducing building logs.

Reply:
So, the flow of what's happening, from further inspection (which lines up with @chengcyber 's comments above):
1. Due to lack of normalized newlines in the data chunks read by the `fs.createReadStream(this._logPath)` in `OperationMetadataManager#tryRestoreAsync`, `DiscardStdoutTransform` throws.
1. This exits the read loop without cleaning up the `Readable` stream, leaking a file handle for `all.log`.
1. Unwinding the stack, this also exits `beforeExecuteOperation` without setting `cacheRestored`.
1. Then control proceeds to the catch block in `OperationExecutionRecord#executeAsync`, which invokes the `onResult` callback.
1. `onResult` calls `CacheableOperationPlugin`'s `afterExecuteOperation` tap.
2. At this point, because Cobuilds have added functionality for recording cache entries in error state, and the `cacheRestored` flag was not set, it calls `OperationMetadataManager#saveAsync`.
3. `OperationMetadataManager#saveAsync` tries to copy to `all.log`, which is locked open for reading by the leaked read stream in `OperationMetadataManager#tryRestoreAsync`, and therefore throws `EPERM`.